### PR TITLE
Add ClusterVirtualMachineImage type

### PR DESCRIPTION
**Pull request by cyclexuxu**
Add ClusterVirtualMachineImage CRD type.

### CODE REVIEW

File: api/v1alpha1/virtualmachineimage_types.go [hunk 1]
```diff
@@ -69,7 +70,7 @@ type VirtualMachineImageSpec struct {
 	ImageID string `json:"imageID"`
 
 	// ProviderRef is a reference to a content provider object that describes a provider.
-	ProviderRef ContentProviderReference `json:"providerRef"`
+	ProviderRef corev1.TypedLocalObjectReference `json:"providerRef"`
```
Comment (dilyar85) [inline api/v1alpha1/virtualmachineimage_types.go:73]: Can you update the comment above to reflect the change here?

File: api/v1alpha1/virtualmachineimage_types.go [hunk 2]
```diff
@@ -115,6 +116,10 @@ type VirtualMachineImageStatus struct {
 	// is supported or image is supported by VMService
 	// +optional
 	Conditions []Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+	// ContentLibraryRef is a reference to the ContentLibrary object
+	// +optional
+	ContentLibraryRef corev1.TypedLocalObjectReference `json:"contentLibraryRef,omitempty"`
```
Comment (dilyar85) [inline api/v1alpha1/virtualmachineimage_types.go:122]: I would rename the description comment to something like `ContentLibraryRef is a reference to the source ContentLibrary/ClusterContentLibrary resource`. Also, I believe this should be a required field same as the `ProviderRef` above.

**Review by dilyar85 — CHANGES_REQUESTED** [review]

**Comment by dilyar85** [comment]
Great job @cyclexuxu! I left some review comments above and overall it looks good to me. Can you please rename the new file to `clustervirtualmachineimage_types.go` (in plural) to keep it consistent with other types? Thanks.

File: api/v1alpha1/clustervirtualmachineimage_type.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+func (clusterVirtualMachineImage *ClusterVirtualMachineImage) GetConditions() Conditions {
+	return clusterVirtualMachineImage.Status.Conditions
+}
+
+func (clusterVirtualMachineImage *ClusterVirtualMachineImage) SetConditions(conditions Conditions) {
+	clusterVirtualMachineImage.Status.Conditions = conditions
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,shortName=vmimage
```
[the remark this answers is not in the document]
Comment (cyclexuxu) [reply api/v1alpha1/clustervirtualmachineimage_type.go]: Thanks for pointing out. Changed the `shortName` as suggested.

File: api/v1alpha1/clustervirtualmachineimage_type.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ContentSourceName",type="string",JSONPath=".spec.providerRef.name"
```
[the remark this answers is not in the document]
Comment (cyclexuxu) [reply api/v1alpha1/clustervirtualmachineimage_type.go]: Change the column name to `ContentLibraryName `. About the value, per discussion on slack, the introduced `contentLibraryRef` was in status, I think the value should be `.status.ContentLibraryRef.name`

File: api/v1alpha1/virtualmachineimage_types.go (hunk 1 above)
Comment (cyclexuxu) [reply api/v1alpha1/virtualmachineimage_types.go:73]: Updated the comment to `ProviderRef is a reference to the source ContentLibraryItem/ClusterContentLibraryItem resource.`

File: api/v1alpha1/virtualmachineimage_types.go (hunk 2 above)
Comment (cyclexuxu) [reply api/v1alpha1/virtualmachineimage_types.go]: Thanks. Renamed the description comment and eliminated the optional comment.

File: api/v1alpha1/clustervirtualmachineimage_types.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+func (clusterVirtualMachineImage *ClusterVirtualMachineImage) SetConditions(conditions Conditions) {
+	clusterVirtualMachineImage.Status.Conditions = conditions
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,shortName=cvmimage;clustervmimage
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ContentLibraryName",type="string",JSONPath=".status.ContentLibraryRef.name"
```
Comment (dilyar85) [inline api/v1alpha1/clustervirtualmachineimage_types.go:22]: @cyclexuxu I might have given the wrong example in my previous review. This should be `.status.contentLibraryRef.name` (lowercase) to display that field properly.

File: api/v1alpha1/virtualmachineimage_types.go [hunk 6]
```diff
@@ -129,7 +133,7 @@ func (vmImage *VirtualMachineImage) SetConditions(conditions Conditions) {
 // +kubebuilder:resource:scope=Cluster,shortName=vmimage
 // +kubebuilder:storageversion
 // +kubebuilder:subresource:status
-// +kubebuilder:printcolumn:name="ContentSourceName",type="string",JSONPath=".spec.providerRef.name"
+// +kubebuilder:printcolumn:name="ContentLibraryName",type="string",JSONPath=".status.ContentLibraryRef.name"
```
Comment (dilyar85) [inline api/v1alpha1/virtualmachineimage_types.go:136]: Ditto

**Review by dilyar85 — APPROVED** [review]
LGTM besides the typos in printcolumn. Please address them before merging this PR.

File: api/v1alpha1/virtualmachineimage_types.go (hunk 6 above)
Comment (cyclexuxu) [reply api/v1alpha1/virtualmachineimage_types.go]: Thanks, fixed typo.

**Review by akutz — APPROVED** [review]
LGTM, thank you!

**Review by zyiyi11 — COMMENTED** [review]
oh I needed to submit...

File: api/v1alpha1/clustervirtualmachineimage_types.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+// +kubebuilder:printcolumn:name="Version",type="string",JSONPath=".spec.productInfo.version"
+// +kubebuilder:printcolumn:name="OsType",type="string",JSONPath=".spec.osInfo.type"
+// +kubebuilder:printcolumn:name="Format",type="string",JSONPath=".spec.type"
+// +kubebuilder:printcolumn:name="ImageSupported",type="boolean",priority=1,JSONPath=".status.imageSupported"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+
+// ClusterVirtualMachineImage is the schema for the clustervirtualmachineimage API
+// A ClusterVirtualMachineImage represents the desired specification and the observed status of a
```
[the remark this answers is not in the document]
Comment (cyclexuxu) [reply api/v1alpha1/clustervirtualmachineimage_types.go]: Thanks. As of now, I think we can depend on the `Cluster` prefix to indicate this is cluster-scoped. If caused confusion, we can update later. 
Yes, ended up with including the changes inside the existing `virtualmachineimage_types.go` file.

**Pull request merged**